updatehub-sdk/src/listener.rs

```diff
@@ -0,0 +1,145 @@
+// Copyright (C) 2018, 2019, 2020 O.S. Systems Sofware LTDA
+//
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{Error, Result};
+use async_ctrlc::CtrlC;
+use async_std::{
+    io::BufReader,
+    os::unix::net::{UnixListener, UnixStream},
+    prelude::*,
+};
+use slog::*;
+use std::{collections::HashMap, net::Shutdown, process};
+
+const SDK_TRIGGER_FILENAME: &str =
+    "/usr/share/updatehub/state-change-callbacks.d/10-updatehub-sdk-statechange-trigger";
+const SOCKET_PATH: &str = "/run/updatehub-statechange.sock";
+
+type StateChangeCallback = Box<dyn Fn(State)>;
+type ErrorCallback = Box<dyn Fn(String)>;
+
+#[derive(Default)]
+pub struct StateChangeListener {
+    listeners: HashMap<StateID, Vec<StateChangeCallback>>,
+    error_handlers: Vec<ErrorCallback>,
+}
+
+#[derive(PartialEq, Eq, Hash)]
+pub enum StateID {
+    Download,
+    Install,
+    Reboot,
+    Error,
+}
+
+pub struct State {
+    pub id: StateID,
+    pub conn: UnixStream,
+}
+
+impl State {
+    pub fn cancel(&mut self) {
+        self.conn.write_all(b"cancel");
+    }
+
+    pub fn proceed(&self) {}
```

Comment:
Please add a comment as done in the Python SDK explaining this does nothing for a reason.